2025/2026 SHIPP update

rtesra opened the pull request:
**Updates to SHIPP calculations after 2025/2026 survey update:**
* Add fix to avoid new infections among KP exceeding total new infections in a given age/district
* Add fix to avoid PSE for KPs exceeding size of non-regular partners group
* Ensure Naomi T3 pulled in as "current estimates" for countries that used custom Naomi version with two household surveys surveys (MWI, ZAF)
* Year for sexual risk behaviour survey results now pulled in from "Model inputs" tab in the SHIPP workbook template. 
  Year set to year of most recent survey with sexual risk behaviour data and to 2018 for countries where
  most recent sexual risk behaviour survey is older that 2018.
* Add option to produce SHIPP without scaling to national envelope of KP PSE and new infections from Goals.

**Updates to SHIPP calculations to separate KPs from elevated risk group:**

Adapted `shipp_combine_cats_female()` to replace 'elevrisk' group with 'sexnonkp and 'kp' group:
* Calculate new variables for non-KP risk group (combining sexcohab and sexnonreg)
* No need to create new KP group as single female KP group 
* Corrected denominator error - to check with @krisher1 
* Calculated low, mod, high and v.high incidence risk groups for:
  - Non-KP total: sexcohab + sexnonreg
  - Non-KP: One cohabiting partner (sexcohab)
  - Non-KP: Non-regular partner
  - KP: FSW (sexpaid12m)
  
 Adapted `shipp_combine_cats_male()` to replace 'elevrisk' group with 'sexnonkp and 'kp' group:
* Calculate new variables for non-KP risk group (combining sexcohab and sexnonreg)
* Calculate new variables for KP risk group (combining pwid and msm)

* Calculated low, mod, high and v.high incidence risk groups for:
  - Non-KP total: sexcohab + sexnonreg
  - Non-KP: One cohabiting partner (sexcohab)
  - Non-KP: Non-regular partner
  - KP: PWID + MSM

krisher1 committed:
Update SHIPP code

Add fixes for:
-KP new infections exceeding new infections in a given age/district
-PSE for KPs exceeding size of non-regular partners group
-survey_year included in call

rtesra committed:
bump version

rtesra committed:
Merge branch 'master' into 2025-shipp-update

rtesra committed:
fix spelling

rtesra committed:
spelling error

rtesra committed:
Separate KP from elevated risk group

rtesra committed:
replace empty columns in male PSE with placeholder names to ensure consistent formatting

rtesra committed:
update survey year to read automatically in internal calculations

jeffeaton committed:
Edits to NEWS.md for clarity

jeffeaton commented on NEWS.md:
```diff
@@ -1,3 +1,12 @@
+* Add fix to avoid PSE for KPs exceeding size of non-regular partners group
+* Ensure Naomi T3 pulled in as "current estimates" for countries that used custom Naomi version with two household surveys surveys (MWI, ZAF)
+* Year for sexual risk behaviour survey results now pulled in from "Model inputs" tab in the SHIPP workbook template. 
+  Year set to year of most recent survey with sexual risk behaviour data and to 2018 for countries where
+  most recent sexual risk behaviour survey is older that 2018.
```
What is the rationale for specifying this as 2018 for earlier surveys rather than the year of the survey?  [Not suggesting a change, just a question for interest and clarification.]

jeffeaton committed:
update white space to match @r-ash style preference

jeffeaton commented on R/shipp.R:
```diff
@@ -1140,7 +1154,7 @@ shipp_calculate_incidence_female <- function(naomi_output,
                                             female_srb,
                                             female_logit_prevalence,
-                                            survey_year = 2018,
+                                            survey_year = 2022,
```
The NEWS.md refers to default survey year being set to 2018, but here in the code 2018 is changed to 2022. Is that referring to the same thing, or is this different from the changes about default year referenced in the NEWS bullet?

jeffeaton committed:
update white space in condition to @r-ash style preference

jeffeaton commented on R/shipp.R:
```diff
@@ -1390,7 +1443,7 @@ shipp_calculate_incidence_male <- function(naomi_output,
                                           male_srb,
                                           male_logit_prevalence,
-                                          survey_year = 2018,
+                                          survey_year = 2022,
```
Same as previous question about 2018 vs. 2022 in NEWS

jeffeaton committed:
update condition white space formatting to @r-ash style preference

jeffeaton committed:
formatting: add white spaces round operators

jeffeaton commented on R/shipp.R:
```diff
@@ -1695,43 +1793,51 @@ shipp_reformat_output <- function(shipp_all) {
+      susceptible_sexnonkp= susceptible_sexnonreg + susceptible_sexcohab,
+      infections_sexnonkp = infections_sexnonreg + infections_sexcohab,
+      incidence_sexnonkp = ((incidence_sexnonreg * susceptible_sexnonreg) +
+                              (incidence_sexcohab * susceptible_sexcohab)) / susceptible_sexnonkp,
```
Could this be equivalently calculated as
```
incidence_sexnonkp = infections_sexnonkp / susceptible_sexnonkp
```
?

Might prefer that for clarity if that's correct

jeffeaton commented on R/shipp.R:
```diff
@@ -1695,43 +1793,51 @@ shipp_reformat_output <- function(shipp_all) {
+                                susceptible_sexnonreg, 0.0),
+      pop_high_inc_sexnonreg = ifelse(incidence_sexnonreg >= 0.005 & incidence_sexnonreg < 0.02,
+                                 susceptible_sexnonreg, 0.0),
+      pop_vhigh_inc_sexnonreg = ifelse(incidence_sexnonreg >= 0.02, susceptible_sexnonreg, 0.0)
```
What is the source and importance of the 0.002, 0.005, and 0.02 magic numbers in the code block below? Are those the incidence threshold categories specified by UNAIDS?

Can we add a comment at the top of this block describing what those numbers are?
And perhaps ideally code those once to variables and reuse the variable name throughout the code below.

jeffeaton commented on R/shipp.R:
```diff
@@ -1779,41 +1913,55 @@ shipp_combine_cats_male <- function(age_filter, shipp, naomi_output) {
+      susceptible_sexnonkp= susceptible_sexnonreg + susceptible_sexcohab,
+      infections_sexnonkp = infections_sexnonreg + infections_sexcohab,
+      incidence_sexnonkp = ((incidence_sexnonreg * susceptible_sexnonreg) +
+                              (incidence_sexcohab* susceptible_sexcohab))/susceptible_sexnonkp,
```
See comment above -- about whether this can be simplified using the `infections_sexnonkp` created on the previous line.

jeffeaton commented on R/shipp.R:
```diff
@@ -1779,41 +1913,55 @@ shipp_combine_cats_male <- function(age_filter, shipp, naomi_output) {
+
+      # Create new variables for population sizes by incidence category
+
+      # Non-KP total: sexcohab + sexnonreg
```
See comment above about changing magic numbers to variable.

jeffeaton reviewed (commented):
Thanks very much -- I've done a couple of formatting PRs, mostly for standardizing white space around operators.

See a few questions and comments; nothing substantive.

rtesra committed:
Update shipp.R

Update to sum pop groups by risk category by 5-year age band

rtesra committed:
Merge branch '2025-shipp-update' of https://github.com/mrc-ide/naomi.utils into 2025-shipp-update

rtesra commented on R/shipp.R:
```diff
@@ -1369,7 +1368,7 @@ shipp_calculate_incidence_female <- function(naomi_output,
                     prev_sexnonreg = plhiv_sexnonreg/(susceptible_sexnonreg + plhiv_sexnonreg),
                     prev_sexpaid12m = plhiv_sexpaid12m/(susceptible_sexpaid12m + plhiv_sexpaid12m),
-                    rr_sexpaid12m = NA)
+                    rr_sexpaid12m = incidence_sexpaid12m/incidence_sexcohab)
```
@krisher1 RR calculated here for combined age categories

rtesra commented on R/shipp.R:
```diff
@@ -1793,11 +1792,24 @@ shipp_reformat_output <- function(shipp_all) {
 
 shipp_combine_cats_female <- function(age_filter, shipp, naomi_output) {
```
@krisher1 Updated to sum population groups from 5-year age bands

rtesra commented on R/shipp.R:
```diff
@@ -1910,11 +1953,25 @@ shipp_combine_cats_female <- function(age_filter, shipp, naomi_output) {
 
 shipp_combine_cats_male <- function(age_filter, shipp, naomi_output) {
```
@krisher1 Updated to sum population groups from 5-year age bands

rtesra committed:
Update shipp.R

Updated risk population sums

rtesra committed:
Update naomi outputs tab name

rtesra committed:
Ensure numeric column in naomi outputs are written out as numbers

krisher1 committed:
Add option to not use Goals

-Input consensus_est = NA if not using Goals or KP Workbooks for the KP new infections

rtesra committed:
Merge branch 'master' into 2025-shipp-update

rtesra committed:
Merge branch '2025-shipp-update' of https://github.com/mrc-ide/naomi.utils into 2025-shipp-update

rtesra committed:
Update tests

rtesra committed:
Update functions so that only FSW newinfections are not scaled to match Goals if desired

rtesra committed:
bump version
